waveorder/filter.py

```diff
@@ -0,0 +1,202 @@
+        sample spacing of i_input_array.
+
+        Leading dimensions are the input and output dimensions.
+        io_filter_bank.shape[:2] == (num_input_channels, num_output_channels)
+
+        Trailing dimensions are spatial frequency dimensions.
+        io_filter_bank.shape[2:] == (Z', Y', X') or (Y', X')
+
+    i_input_array : torch.Tensor
+        The real-valued input array with sample spacing dx to be filtered.
+
+        Leading dimension is the input dimension, matching the filter bank.
+        i_input_array.shape[0] == i
+
+        Trailing dimensions are spatial dimensions.
+        i_input_array.shape[1:] == (Z, Y, X) or (Y, X)
+
+    Returns
+    -------
+    torch.Tensor
+        The filtered real-valued output array with shape
+        (num_output_channels, Z, Y, X) or (num_output_channels, Y, X).
+
+    """
+
+    # Ensure all dimensions of transfer_function are smaller than or equal to input_array
+    if any(
+        t > i
+        for t, i in zip(io_filter_bank.shape[2:], i_input_array.shape[1:])
+    ):
+        raise ValueError(
+            "All spatial dimensions of io_filter_bank must be <= i_input_array."
+        )
+
+    # Ensure the number of spatial dimensions match
+    if io_filter_bank.ndim - i_input_array.ndim != 1:
+        raise ValueError(
+            "io_filter_bank and i_input_array must have the same number of spatial dimensions."
+        )
+
+    # Ensure the input dimensions match
+    if io_filter_bank.shape[0] != i_input_array.shape[0]:
+        raise ValueError(
+            "io_filter_bank.shape[0] and i_input_array.shape[0] must be the same."
+        )
+
+    num_input_channels, num_output_channels = io_filter_bank.shape[:2]
+    spatial_dims = io_filter_bank.shape[2:]
+
+    # Pad input_array until each dimension is divisible by transfer_function
+    pad_sizes = [
+        (0, (t - (i % t)) % t)
+        for t, i in zip(
+            io_filter_bank.shape[2:][::-1], i_input_array.shape[1:][::-1]
+        )
+    ]
+    flat_pad_sizes = list(itertools.chain(*pad_sizes))
+    padded_input_array = torch.nn.functional.pad(i_input_array, flat_pad_sizes)
+
+    # Apply the transfer function in the frequency domain
+    fft_dims = [d for d in range(1, i_input_array.ndim)]
+    padded_input_spectrum = torch.fft.fftn(padded_input_array, dim=fft_dims)
+
+    # Matrix-vector multiplication over f
+    # If this is a bottleneck, consider extending `stretched_multiply` to
+    # a `stretched_matrix_multiply` that uses an call like
+    # torch.einsum('io..., i... -> o...', io_filter_bank, padded_input_spectrum)
+    padded_output_spectrum = torch.zeros(
+        (num_output_channels,) + spatial_dims,
+        dtype=padded_input_spectrum.dtype,
+    )
+    for input_channel_idx in range(num_input_channels):
+        for output_channel_idx in range(num_output_channels):
+            padded_output_spectrum[output_channel_idx] += stretched_multiply(
```

Comment:
Torch has some special accelerated calls for block-wise multiply-add. But maybe speed is not a bottleneck here.

Reply:
In my latest tests this is currently not a bottleneck (we make at most 3 input x 4 output = 12 trips through this loop), but it could become a bottleneck in the future. 

Still I'd appreciate a link to the block-wise multiply-adds that you have in mind. I'm seeing this [`bmm`](https://pytorch.org/docs/stable/generated/torch.bmm.html), but that's not quite what I'd need here.

Reply:
I was thinking of something like [baddmm](https://pytorch.org/docs/stable/generated/torch.baddbmm.html#torch.baddbmm) with reshaping. But I know realize that the 'blocks' here are super small (2x2?) so the memory access pattern is not really 'blocky'. The current approach is similar to pixel-[shuffle](https://pytorch.org/docs/stable/generated/torch.nn.functional.pixel_shuffle.html#torch.nn.functional.pixel_shuffle)/[unshuffle](https://pytorch.org/docs/stable/generated/torch.nn.functional.pixel_unshuffle.html#torch.nn.functional.pixel_unshuffle) for stretched-multiply followed by a point-wise convolution done here.

Reply:
(And point-wise convolution is really just a different way to spell 'nn.Linear')

Reply:
Can it be done via pixel-shuffle -> linear -> pixel-unshuffle?

Reply:
Thanks @ziw-liu. After a quick in-person chat we agreed that this is likely premature optimization, but we agreed that there is likely additional optimization available here.  

I've added a comment that points to these functions so that we can optimize further if we find that this is a bottleneck.